File: price_tracker.py
import asyncio
import csv
import os
from datetime import datetime
from playwright.async_api import async_playwright

# Function to extract text safely with a timeout
async def get_text(page, selector, default="Not found"):
    try:
        element = await page.wait_for_selector(selector, timeout=5000)
        return await element.text_content() if element else default
    except:
        return default

# Function to get the discounted price with fallbacks
async def get_discounted_price(page):
    discounted_price = await get_text(page, ".apexPriceToPay span.a-offscreen")
    if discounted_price == "Not found":
        discounted_price = await get_text(page, ".a-price .a-offscreen")  # Fallback

    return discounted_price.strip()

# Function to scrape individual product data
async def scrape_amazon_product(asin):
    url = f"https://www.amazon.in/dp/{asin}"

    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=True)
        context = await browser.new_context()
        page = await context.new_page()

        try:
            await page.goto(url, timeout=15000)

            # Extract stock status
            stock_status = await get_text(page, "#availability span")

            # Skip extracting if stock status is "Not found"
            if stock_status == "Not found":
                print(f"⚠️ Skipping {asin}: Stock status could not be determined.")
                await browser.close()
                return None

            # Check if the product is out of stock
            if "out of stock" in stock_status.lower():
                print(f"⚠️ Skipping {asin}: Out of stock")
                await browser.close()
                return None

            # Extract discounted price
            discounted_price = await get_discounted_price(page)
            
            # Check if the price was found
            if discounted_price == "Not found":
                print(f"⚠️ Skipping {asin}: Discounted price could not be found.")
                await browser.close()
                return None

            await browser.close()
            return discounted_price

        except Exception as e:
            await browser.close()
            print(f"⚠️ Error fetching product {asin}: {e}")
            return None

# Read ASINs from a text file
def get_asins_from_file(filename):
    try:
        with open(filename, "r") as file:
            return [line.strip() for line in file.readlines()]
    except FileNotFoundError:
        print(f"❌ Error: File '{filename}' not found.")
        return []

# Function to update CSV file with new prices and a timestamp header
def update_csv(prices, filename="amazon_price_tracking.csv"):
    timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    file_exists = os.path.isfile(filename)

    if not file_exists:
        with open(filename, mode="w", newline="", encoding="utf-8") as file:
            writer = csv.writer(file)
            header = ["ASIN", timestamp]
            writer.writerow(header)
    else:
        with open(filename, mode="r", newline="", encoding="utf-8") as file:
            reader = csv.reader(file)
            data = list(reader)
        
        # Create a dictionary from existing CSV data
        data_dict = {row[0]: row for row in data}
        
        # Add new timestamp header if needed
        if len(data[0]) == len(data_dict[next(iter(data_dict))]):
            for row in data:
                row.append("")
        
        data[0].append(timestamp)

    # Update the dictionary with new prices
    for asin, price in prices.items():
        if asin in data_dict:
            data_dict[asin].append(price)
        else:
            data_dict[asin] = [asin] + [""] * (len(data[0]) - 2) + [price]

    # Write updated data back to CSV
    with open(filename, mode="w", newline="", encoding="utf-8") as file:
        writer = csv.writer(file)
        for row in data_dict.values():
            writer.writerow(row)

# Main function to scrape all products from the list of ASINs
async def scrape_amazon_products():
    filename = "C:/Users/project/Downloads/url500.txt"
    asin_list = get_asins_from_file(filename)

    if not asin_list:
        print("❌ No ASINs found. Exiting...")
        return

    prices = {}
    for asin in asin_list:
        discounted_price = await scrape_amazon_product(asin)
        if discounted_price:
            prices[asin] = discounted_price
            print(f"✅ Price fetched for {asin}")
        print("-" * 80)

    # Update the CSV file with the new prices
    update_csv(prices)

# Execute script
if __name__ == "__main__":
    asyncio.run(scrape_amazon_products())

File: price_tracker.js
const fs = require('fs');
const puppeteer = require('puppeteer'); // Assuming you're using Puppeteer for scraping

async function getTextContent(page, selector) {
    const element = await page.$(selector);
    if (element) {
        return await page.evaluate(el => el.textContent.trim(), element);
    }
    return null; // Return null if the element is not found
}

async function getPrice(page, asin) {
    try {
        const url = `https://www.amazon.com/dp/${asin}`;
        await page.goto(url, { waitUntil: 'domcontentloaded' });

        // Check if the page title contains 'Amazon'
        const pageTitle = await page.title();
        if (!pageTitle.includes("Amazon")) {
            console.log(`Failed to load the product page for ASIN: ${asin}`);
            return 'Error loading page';
        }

        // Wait for the price element to appear
        await page.waitForSelector('.a-price-whole', { timeout: 5000 });

        // Fetch the price components from the page
        const discountedPrice = await getTextContent(page, ".a-price-whole");
        const discountedPriceFraction = await getTextContent(page, ".a-price-fraction");

        if (discountedPrice && discountedPriceFraction) {
            const fullPrice = `${discountedPrice}.${discountedPriceFraction}`;
            console.log(`Price for ASIN ${asin}: $${fullPrice}`);
            return fullPrice;
        } else {
            console.log(`Price not found for ASIN ${asin}`);
            return 'Price not available';
        }
    } catch (error) {
        console.error(`Error fetching price for ASIN ${asin}:`, error);
        return 'Error fetching price';
    }
}

async function updatePricesFromFile() {
    // Path to the .txt file containing ASINs
    const filePath = 'C:/Users/vishn/Documents/url500.txt';

    // Read ASINs from the .txt file
    const asins = fs.readFileSync(filePath, 'utf-8').split('\n').map(line => line.trim()).filter(line => line);

    // Launch Puppeteer browser instance
    const browser = await puppeteer.launch();
    const page = await browser.newPage();

    // Loop through each ASIN in the text file
    for (let asin of asins) {
        console.log(`Fetching price for ASIN: ${asin}...`);

        // Fetch the price using the getPrice function
        const price = await getPrice(page, asin);

        // Display the price in the console
        console.log(`Price for ASIN ${asin}: ${price}\n`);
    }

    // Close the Puppeteer browser
    await browser.close();
}

// Call the function to update the prices
updatePricesFromFile().catch(console.error);
